cli/crates/gateway/src/context.rs

```diff
@@ -0,0 +1,29 @@
+use futures_util::future::BoxFuture;
+use http::HeaderMap;
+use tokio::{sync::Mutex, task::JoinHandle};
+
+pub struct Context {
+    pub(crate) ray_id: String,
+    pub(crate) x_api_key_header: Option<String>,
+    pub(crate) authorization_header: Option<String>,
+    pub(crate) headers: HeaderMap,
+    // TODO: or use a queue?
+    pub(crate) join_handles: Mutex<Vec<JoinHandle<()>>>,
+}
+
+#[async_trait::async_trait]
+impl gateway_core::RequestContext for Context {
+    fn ray_id(&self) -> &str {
+        &self.ray_id
+    }
+
+    async fn wait_until(&self, fut: BoxFuture<'static, ()>) {
+        // FIXME: should be propagated.
+        let handle = tokio::runtime::Handle::current();
+        self.join_handles.lock().await.push(handle.spawn(fut));
+    }
```

Comment:
I'm not sure that this should spawn.  Don't think it was the behaviour in the existing gateway anyway - pretty sure it waits until the request is finished before running the `wait_until` futures.